packages/react-data-grid/src/utils/viewportUtils.ts

```diff
@@ -60,21 +60,29 @@ export function getHorizontalRangeToRender<R>({
   let { viewportWidth } = columnMetrics;
 
   let remainingScroll = scrollLeft;
```

Comment:
```suggestion
  // Get valid left scroll position.
  // When we're scrolled all the way to the right and we remove columns, the scrollLeft value will be outdated,
  // and greater than what's possible, so we need to handle that case here.
  let remainingScroll = scrollLeft + viewportWidth > totalColumnWidth
    ? totalColumnWidth - viewportWidth
    : scrollLeft;
```

I've tested this locally, and I think this is the only change we need, you can revert the other changes and apply this, unless I've missed something.